### PR TITLE
Feature/preprocessing additional checks

### PR DESCRIPTION
**Pull request by johannmartin**
# Description

Adding additional checks on preprocessing for SmartPredictor and allowing more preprocessing options.
- Authorize OneHotEncoders for Category_encoders preprocessing (#58 )
- Authorize OneHotEncoders for ColumnTransformers preprocessing (#64 )
- Authorize the 'drop' options for ColumnTransformers preprocessing
- Add additional checks on the preprocessing to make sure the output fit the expected input of the model (#46 )
- Add additional checks for list of preprocessing 

## Type of change

- [ ] New feature (non-breaking change which adds functionality or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

- [ ] Test with the SmartPredictor on jupyter notebook
- [ ] Add additional Unit tests on each method added 

**Test Configuration**:
* OS: Linux
* Python version: 3.6/3.7/3.8
* Shapash version: 1.1.0

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: shapash/explainer/smart_predictor.py [hunk 1]
```diff
@@ -623,22 +635,41 @@ def summarize(self):
...
-            rank_contributions(
-                self.data["contributions"],
-                self.data["x_postprocessed"]
+        if self._drop_option is not None:
+            self.summary = assign_contributions(
+                rank_contributions(
+                    self.data["contributions"],
+                    self.data["x_postprocessed"][self.data["x_postprocessed"].\
+                                                columns.difference(self._drop_option["features_to_drop"])]
+                )
```
Comment (ThomasBouche) [inline shapash/explainer/smart_predictor.py:644]: Can you make the code clearer

File: shapash/utils/check.py [hunk 2]
```diff
@@ -218,27 +218,49 @@ def check_consistency_model_features(features_dict, model, columns_dict, feature
...
+        if model_expected != len(feature_encoded):
+            raise ValueError("""
+                Number of features returned by the list preprocessing doesn't
+                match the model's expected features.
+                             """)
+
+    elif preprocessing is not None and str(type(preprocessing)) != "<class 'dict'>":
+        raise ValueError("{0} type is not supported in SmartPredictor".format(str(type(preprocessing))))
 
     if postprocessing:
```
Comment (ThomasBouche) [inline shapash/utils/check.py:265]: Not missing an argument ?

File: shapash/utils/check.py [hunk 3]
```diff
@@ -4,13 +4,13 @@
```
Comment (ThomasBouche) [inline shapash/utils/check.py:4]: Maybe check_consistency_model_features start to be difficult to read between specific encoders and specif models

File: shapash/utils/check.py [hunk 4]
```diff
@@ -218,27 +218,49 @@ def check_consistency_model_features(features_dict, model, columns_dict, feature
...
             raise ValueError("All features of preprocessing must be in columns_dict")
 
+    if str(type(preprocessing)) in columntransformer:
+        for encoding in preprocessing.transformers_:
+            ct_encoding = encoding[1]
+            if (str(type(ct_encoding)) not in supported_category_encoder) \
+                    and (str(type(ct_encoding)) not in supported_sklearn):
+                if str(type(ct_encoding)) != "<class 'str'>":
+                    raise ValueError("{0} used in ColumnTransformers isn't supported.".format(str(type(ct_encoding))))
+
```
Comment (SebastienBidault) [inline shapash/utils/check.py:232]: already check in check_transformers

File: shapash/utils/check.py [hunk 5]
```diff
@@ -218,27 +218,49 @@ def check_consistency_model_features(features_dict, model, columns_dict, feature
...
+                                             """)
+
+    elif str(type(preprocessing)) in columntransformer:
+        feature_encoded = get_feature_names(preprocessing)
+        if model_expected != len(feature_encoded):
+            raise ValueError("""
+                Number of features returned by the ColumnTransformer preprocessing doesn't
+                match the model's expected features.
+                             """)
+
```
Comment (SebastienBidault) [inline shapash/utils/check.py:253]: Maybe use preprocessing_tolist to avoid multiply check

File: shapash/utils/check.py [hunk 6]
```diff
@@ -250,21 +272,52 @@ def check_consistency_model_features(features_dict, model, columns_dict, feature
...
+        Dictionary mapping technical feature names to domain names.
+    Returns
+    -------
+    None, dict
+        None if there isn't drop options in ColumnTransformer otherwise dict of informations to adapt.
     """
+    feature_to_drop = list()
     if preprocessing is not None:
         list_encoding = preprocessing_tolist(preprocessing)
+
```
Comment (SebastienBidault) [inline shapash/utils/check.py:296]: maybe try using preprocessing_tolist once at the init part ?

File: shapash/utils/columntransformer_backend.py [hunk 7]
```diff
@@ -293,4 +294,100 @@ def transform_ct(x_in, model, encoding):
...
+
+    for enc in list_preprocessing:
+        if str(type(enc)) in supported_category_encoder:
+            feature_expected = enc.feature_names
+
+        elif str(type(enc)) in columntransformer:
+            feature_expected = get_feature_names(enc)
+
+        elif str(type(enc)) in ("<class 'dict'>", "<class 'list'>"):
+            feature_expected = feature_expected
```
Comment (SebastienBidault) [inline shapash/utils/columntransformer_backend.py:390]: dict is transform as list in preprocessing_tolist

File: tests/unit_tests/utils/test_check.py [hunk 8]
```diff
@@ -388,6 +388,88 @@ def test_check_consistency_model_features_4(self):
...
+        features_dict = None
+        columns_dict = {i: features for i, features in enumerate(train.columns)}
+        features_types = {features: str(train[features].dtypes) for features in train.columns}
+        mask_params = None
+
+        enc = ColumnTransformer(
+            transformers=[
+                ('onehot_ce', ce.OrdinalEncoder(), ['city', 'state']),
+                ('onehot_skp', skp.OrdinalEncoder(), ['city', 'state'])
+            ],
```
Comment (SebastienBidault) [inline tests/unit_tests/utils/test_check.py:409]: rename onehot ordinal

File: tests/unit_tests/utils/test_check.py [hunk 9]
```diff
@@ -388,6 +388,88 @@ def test_check_consistency_model_features_4(self):
...
+                  'data_type': 'object'}
+
+        dict_5 = {'col': 'city',
+                  'mapping': pd.Series(data=[1, 2],
+                                       index=['chicago', 'paris']),
+                  'data_type': 'object'}
+
+        enc_4 = [enc_3, dict_4]
+
+        enc_5 = [enc_3, [dict_4]]
```
Comment (SebastienBidault) [inline tests/unit_tests/utils/test_check.py:456]: add dict_5 after dict_4

File: tests/unit_tests/utils/test_columntransformer_backend.py [hunk 10]
```diff
@@ -959,4 +960,60 @@ def test_transform_ct_8(self):
...
+        assert all(expected.index == result.index)
+
+    def test_get_feature_names_1(self):
+        """
+        Unit test get_feature_names 1
+        """
+        train = pd.DataFrame({'num1': [0, 1],
+                              'num2': [0, 2],
+                              'other': ['A', 'B']})
+
```
Comment (SebastienBidault) [inline tests/unit_tests/utils/test_columntransformer_backend.py:972]: rename pwer

File: tests/unit_tests/utils/test_columntransformer_backend.py [hunk 11]
```diff
@@ -959,4 +960,60 @@ def test_transform_ct_8(self):
...
+        enc_1 = ColumnTransformer(transformers=[('power', skp.QuantileTransformer(n_quantiles=2), ['num1', 'num2'])],
+                                remainder='drop')
+
+        enc_2 = ColumnTransformer(transformers=[('power', skp.QuantileTransformer(n_quantiles=2), ['num1', 'num2'])],
+                                remainder='passthrough')
+
```
Comment (SebastienBidault) [inline tests/unit_tests/utils/test_columntransformer_backend.py:978]: add one more test where a dummies is done

**Review by ThomasBouche — APPROVED** [review]

**Review by ThomasBouche — APPROVED** [review]

**Review by ThomasBouche — APPROVED** [review]

**Pull request merged**